pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,321 @@
+	"antrea.io/antrea/pkg/util/k8s"
+)
+
+const (
+	defaultClusterHealthzAskInterval = 40 * time.Second
+	// Set resyncPeriod to 0 to disable resyncing.
+	resyncPeriod time.Duration = 0
+	// set default virtual node replicas num of consistent hash
+	defaultVirtualNodeReplicas = 3
+)
+
+// defaultHashFn is crc32.ChecksumIEEE
+var defaultHashFn func(data []byte) uint32
+
+type ClusterInterface interface {
+	Run(stopCh <-chan struct{})
+	ShouldSelect(name string) bool
+	AddEventHandler(handler clusterNodeEventHandler)
+}
+
+type clusterNodeEventHandler func(nodeName string, isJoin bool)
+
+type Cluster struct {
+	bindPort                         int
+	NodeConfig                       *config.NodeConfig
+	nodeInformer                     coreinformers.NodeInformer
+	memberList                       *memberlist.Memberlist
+	memberRWLock                     sync.RWMutex
+	existingMembers                  []string
+	defaultClusterHealthzAskInterval time.Duration
+	conHash                          *consistenthash.Map
+	nodeEventsCh                     chan memberlist.NodeEvent
+	ClusterNodeEventHandlers         []clusterNodeEventHandler
+
+	ipPoolInformer          crdinformers.ExternalIPPoolInformer
+	ipPoolLister            crdlister.ExternalIPPoolLister
+	ipPoolInformerHasSynced cache.InformerSynced
+}
+
+func NewCluster(
+	clusterBindPort int,
+	nodeInformer coreinformers.NodeInformer,
+	nodeConfig *config.NodeConfig,
+	ipPoolInformer crdinformers.ExternalIPPoolInformer,
+) (*Cluster, error) {
+	// The Node join/leave events will be notified via it.
+	nodeEventCh := make(chan memberlist.NodeEvent, 1024)
+	c := &Cluster{
+		bindPort:                         clusterBindPort,
+		nodeInformer:                     nodeInformer,
+		NodeConfig:                       nodeConfig,
+		defaultClusterHealthzAskInterval: defaultClusterHealthzAskInterval,
+		nodeEventsCh:                     nodeEventCh,
+		ipPoolInformer:                   ipPoolInformer,
+		ipPoolLister:                     ipPoolInformer.Lister(),
+		ipPoolInformerHasSynced:          ipPoolInformer.Informer().HasSynced,
+	}
+	c.conHash = newNodeConsistentHashMap()
+
+	bindPort := c.bindPort
+	hostIP := c.NodeConfig.NodeIPAddr.IP
+
+	nodeMember := fmt.Sprintf("%s:%d", hostIP.String(), bindPort)
+
+	klog.V(2).Infof("Add new node: %s", nodeMember)
+
+	conf := memberlist.DefaultLocalConfig()
+	conf.Name = c.NodeConfig.Name
+
+	conf.BindPort = bindPort
+	conf.AdvertisePort = bindPort
+	conf.Events = &memberlist.ChannelEventDelegate{Ch: nodeEventCh}
+
+	klog.V(1).Infof("Memberlist cluster configs: %+v", conf)
+
+	mList, err := memberlist.Create(conf)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create memberlist: %v", err.Error())
+	}
+
+	c.memberList = mList
+	c.existingMembers = []string{nodeMember}
+
+	nodeInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addNodeMemberHandler,
+			UpdateFunc: nil,
```

Comment:
Should it watch update events as well so it can leave the cluster when it's label changes?

Reply:
when updating labels of a node, 
local node Processing：
if the node has joined the cluster, new labeles not match externalIPPool, the node will leave cluster,
if the node has not joined the cluster, new labels match the externalIPPool, the node will join cluster,
other nodeProcessing：
add or delete from existingMember list,  nodeEventsCh will receive a signal and handle nodes consistent hash Map